Add layout creation convenience methods

lukpueh opened the pull request:
**Description of the changes being introduced by the pull request**:
Adds methods to the layout class to ease layout creation.

Here is a list of the newly added methods:
- set_relative_expiration
- get_step_name_list
- get_step_by_name
- remove_step_by_name
- get_inspection_name_list
- get_inspection_by_name
- remove_inspection_by_name
- get_functionary_key_id_list
- add_functionary_key
- add_functionary_key_from_path
- add_functionary_key_from_gpg_keyid
- add_functionary_keys_from_paths
- add_functionary_keys_from_gpg_keyids

Additionally refines and fixes various docstrings, comments, exceptions messages and spacings in the the layout class.

Please merge #180 and #179 first. I'd like to give the docstrings, comments in the parts of the layout module affected by those PRs another pass here.

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

Commit:
Add layout convenience methods

Add methods to the layout class to ease layout creation and
refine various docstrings, comments, exceptions messages and
spacings in the the layout class.

Here is a list of the newly added methods:
- set_relative_experation
- get_step_name_list
- get_step_by_name
- remove_step_by_name
- get_inspection_name_list
- get_inspection_by_name
- remove_inspection_by_name
- get_functionary_key_id_list
- add_functionary_key
- add_functionary_key_from_path
- add_functionary_key_from_gpg_keyid
- add_functionary_keys_from_paths
- add_functionary_keys_from_gpg_keyids

Commit:
Add unittests for new Layout methods

Adds unittests, fixes docstrings and line spacing.

SantiagoTorres commented on in_toto/models/layout.py:
```diff
@@ -87,31 +66,99 @@ class Layout(Signable):
+
+    <Arguments>
+      data:
+              A dictionary containing layout metadata.
```
Can we add "conforming to the in-toto specification" (and probably the version too?)

lukpueh replied:
Will fix with #186.

SantiagoTorres commented on in_toto/models/layout.py:
```diff
@@ -126,12 +173,376 @@ def read(data):
+    securesystemslib.schema.Integer().check_match(years)
+
+    self.expires = (datetime.today() + relativedelta(
+        days=days, months=months, years=years)).strftime("%Y-%m-%dT%H:%M:%SZ")
```
Do we have a schema for date? can we add a guard?

lukpueh replied:
There's a `_validate_expires` method on the Layout, which we could use (it checks schema and date validity). But I'm unit testing `set_relative_expiration` together with `_validate_expires`, so I don't think it is necessary to add a guard here.

SantiagoTorres commented on in_toto/models/layout.py:
```diff
@@ -126,12 +173,376 @@ def read(data):
+    """
+    keyid_list = []
+    for keyid in list(self.keys.keys()):
+      keyid_list.append(keyid)
```
I know this goes against the our guidelines, but a list comprehension here would be way more readable imho.

lukpueh replied:
I guess that's a matter of taste. And while I'm a big fan of list comprehensions, I think we should stick to our guidelines.

vladimir-v-diaz replied:
Why is `keyid_list` needed?  Doesn't the following work?
```Python
return list(self.keys.keys())
```

Note: the list() is needed for Python 2+3 compatibility, otherwise `return self.keys.keys()` should be fine.

lukpueh replied:
Good thinking Vlad! :)

SantiagoTorres commented on tests/models/test_layout.py:
```diff
@@ -19,25 +19,198 @@
+    layout.set_relative_expiration(years=3)
+    layout._validate_expires()
+    layout.set_relative_expiration(days=3, months=2, years=1)
+    layout._validate_expires()
```
can you throw some negative numbers in here to see what happens? :)

lukpueh replied:
Sure, will do.

SantiagoTorres commented on tests/models/test_layout.py:
```diff
@@ -19,25 +19,198 @@
+    layout = Layout(steps=[Step(name=name) for name in names])
+    layout.remove_step_by_name("b")
+    self.assertEqual(len(layout.steps), 2)
+    self.assertTrue("b" not in layout.get_step_name_list())
```
Can you try a double-remove?

lukpueh replied:
Sure. What's expected behavior? Currently it would do nothing.

SantiagoTorres commented on tests/models/test_layout.py:
```diff
@@ -19,25 +19,198 @@
+    self.assertTrue("b" not in layout.get_inspection_name_list())
+
+    with self.assertRaises(securesystemslib.exceptions.FormatError):
+      layout.remove_inspection_by_name(False)
```
I wonder if testing the following sequence would be a good idea:

1. Add step `a`
1. Remove step `a`
1. Add *inspection* `a`
1. Remove inspection `a`

lukpueh replied:
Why exactly? Note, I didn't add `add_{step, inspection}` methods in this PR, which need testing. Currently the remove methods are tested separately like so:
```
for item in [step, inspection]:
    create layout with items a, b, c
    remove item a
    assert that item is not on on layout
```
However, I can add an assert that the item is on the layout before being removed.

SantiagoTorres reviewed (changes requested):
Looking good, but I think it's missing a little bit more tests.

Commit:
Add relative expiration test for dates in the past

Commit:
Update remove_step and remove_inspection tests

- Assert for existence of item to be removed, before removal.
- Try double removal.

Commit:
Use dict builtin for get_functionary_key_id_list

dict.keys() already returns a list of the dict's keys. We don't
have to iterate over that list's items and manually create a new
list.

Commit:
Add artifact rule format check in setters

Commit:
Remove obsolete inpsection attributes/assignments

Inspections inherit name and expected_* fields from
SupplyChainItem, whose constructor also assign the values from
optionally passed kwargs.
This commit removes obsolete re-definitions/assignments from
the inspection class.

Commit:
Clean up docstrings in layout module

lukpueh commented:
Thanks, @SantiagoTorres and @vladimir-v-diaz. I replied to/ addressed your comments. And did some more docstring cleaning. Mind giving it another pass?
